### PR TITLE
[image_picker] Added image_picker_for_windows.

### PR DESCRIPTION
**Pull request by azchohfi**
This PR adds a very simplistic Windows implementation of the image_picker plugin, which uses the file_selector plugin to get images or videos from the "gallery". This initial implementation doesn't yet support ImageSource.camera (falls back to gallery).

Fixes https://github.com/flutter/flutter/issues/92108

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [X] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/image_picker/image_picker_for_windows/AUTHORS
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/AUTHORS:67]: The authors list for a new plugin should only contain Google (the first entry here) and the plugin author (if different).

File: packages/image_picker/image_picker_for_windows/README.md
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/README.md:1]: We should call it `image_picker_windows`, as that's our standard naming. The "_for_web" in some packages was to work around the fact that in some cases a third party published something under the "_web" name before we added web support.

File: packages/image_picker/image_picker_for_windows/README.md
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/README.md:20]: What's the difference here relative to normal `image_picker` code?

File: packages/image_picker/image_picker_for_windows/example/lib/main.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/example/lib/main.dart:12]: We don't use the app-facing platform in implementation examples (there are some legacy violations, but we don't do it any more), as it creates a circular dependency that causes problems for testing changes. This should instead be written directly against `image_picker_platform_interface`.

File: packages/image_picker/image_picker_for_windows/example/lib/main.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/example/lib/main.dart:72]: No need for web-specific code in the Windows example.

File: packages/image_picker/image_picker_for_windows/example/lib/main.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/example/lib/main.dart:244]: Or Android.

File: packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart:14]: This comment seems wrong?

File: packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart:17]: We should add at least webp, gif, tiff (and tif), and maybe apng.

File: packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart:18]: webm, avi, mpeg, mpg

File: packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart:13]: We generally don't include a `Plugin` suffix here.

File: packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart:36]: We generally don't duplicate declaration comments on overriden methods.

I would just put a non-documentation comment (`//`) with the note about the arguments that are currently ignored.

File: packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart:47]: Prefer implementing this by calling `getImage` so we don't have two copies of the implementation.

File: packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/lib/image_picker_for_windows.dart:74]: Same.

File: packages/image_picker/image_picker_for_windows/pubspec.yaml
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/pubspec.yaml:5]: Nit: Let's make it 0.1.0; we usually start with a minor version.

File: packages/image_picker/image_picker_for_windows/pubspec.yaml
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/pubspec.yaml:9]: 2.5 (required for Dart registration)

File: packages/image_picker/image_picker_for_windows/pubspec.yaml
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/pubspec.yaml:22]: This should be `file_selector_platform_interface`; we don't want to pull in the native `file_selector` implementations for other platforms and bloat people's apps unnecessarily.

File: packages/image_picker/image_picker_for_windows/pubspec.yaml
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_windows/pubspec.yaml:24]: This isn't necessary.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Looks good! This will need Dart unit tests that sets a mock `FileSelectorPlatform` implementation as `FileSelectorPlatform.instance` and validates that the expected calls are made. And also that `registerWith` works (as in other Dart plugin implementation tests).

While the logic is currently pretty trivial, having tests in place is important as we change it over time.

File: packages/image_picker/image_picker_for_windows/README.md
Comment (azchohfi) [reply packages/image_picker/image_picker_for_windows/README.md:20]: I just wanted to say that it is not fully featured yet. I'll change comment to make it clearer.

File: packages/image_picker/image_picker_windows/README.md
Comment (azchohfi) [reply packages/image_picker/image_picker_windows/README.md:22]: Agreed. I can just remove the "Use the plugin" section.

File: packages/image_picker/image_picker_windows/example/lib/main.dart
Comment (azchohfi) [reply packages/image_picker/image_picker_windows/example/lib/main.dart:53]: I've tried doing this, but then I get an exception about the file_selector plugin being missing, during runtime. It seems like the FIleSelectorWindows plugin is not registering itself. If I add a:
```dart
FileSelectorWindows.registerWith();
```
at ImagePickerWindows.registerWith() method, then everything works, but it feels like wrong. Am I missing something here?

**Comment by azchohfi** [comment]
Oh, I believe I need to add the image_picker_windows to the exclude_integration_win32.yaml file.

**Comment by azchohfi** [comment]
@stuartmorgan anything else needed here?

**Comment by stuartmorgan** [comment]
Sorry for the slow response, I haven't had much time for reviews because of the build issue with VS 17.1.0.

File: packages/image_picker/image_picker_windows/example/lib/main.dart
Comment (stuartmorgan) [reply packages/image_picker/image_picker_windows/example/lib/main.dart]: There may well be a bug with Dart registration for the edge case of depending directly on a federated implementation (without the app-facing package). I'll investigate tomorrow or Friday.

Comment (stuartmorgan) [reply packages/image_picker/image_picker_windows/example/lib/main.dart]: Okay, I know why this happens now, but I need to think about it more to figure out if it's actually a bug. There are some weird subtleties.

For the purposes of this PR, let's assume it's not a bug. In which case, the way to handle it is that you shouldn't be using `FileSelectorPlatform` anywhere in your implementation, but instead use a `FileSelectorWindows` instance directly.

Comment (stuartmorgan) [reply packages/image_picker/image_picker_windows/example/lib/main.dart]: (Filed https://github.com/flutter/flutter/issues/99083 for the reason this happens)

Comment (azchohfi) [reply packages/image_picker/image_picker_windows/example/lib/main.dart]: Should I just create my own FileSelectorWindows instance, or should I try to get the FileSelectorPlatform.instance first? A very low overhead since the MethodChannel itself is already a const outside the instance anyway, but I wonder how much this can impact other plugins that follow the same pattern.

Comment (stuartmorgan) [reply packages/image_picker/image_picker_windows/example/lib/main.dart]: You can just make one; it should be minimal cost.

> I wonder how much this can impact other plugins that follow the same pattern

Yes, longer term I do think we'll need to fix the underlying issue. (I don't expect this to be a super common pattern though.)

Comment (azchohfi) [reply packages/image_picker/image_picker_windows/example/lib/main.dart]: Ok, tests are passing, so this should be good now.

File: packages/image_picker/image_picker_windows/README.md
Comment (stuartmorgan) [inline packages/image_picker/image_picker_windows/README.md:4]: This should mention that image source is not currently supported.

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:41]: This doesn't do what I expect you intended; this is basically just an alias for the constructor. Assuming you wanted a shared instance you want:
```
static late FileSelectorWindows = FileSelectorWindows();
```

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:48]: This list is missing `source`

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:93]: For `source`, since ignoring it will do completely the wrong thing (as opposed to the other arguments which are less critical), I think you should throw an `UnimplementedError` when it's not the gallery.

(Ideally we would delegate to `camera_windows`, but without the camera controller that would be quite tricky. Longer term we might have to consider some refactoring there to share code.)

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
A few other small things, but otherwise it should be good from my side.

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (azchohfi) [reply packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:41]: Oh, my bad. I want a simple singleton, and I don't see many reasons to late initialize this. I believe this is fine:
```dart
  static final FileSelectorWindows _fileSelectorInstance = FileSelectorWindows();
```

**Comment by azchohfi** [comment]
@stuartmorgan I think I got it all, let me know if there is anything else needed.

File: packages/image_picker/image_picker_windows/README.md
Comment (azchohfi) [reply packages/image_picker/image_picker_windows/README.md:4]: Done.

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (azchohfi) [reply packages/image_picker/image_picker_windows/lib/image_picker_windows.dart]: Done.

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (azchohfi) [reply packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:97]: I've added the UnimplementedError.

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:110]: Nit (here and below): How about "ImageSource.gallery is currently the only supported source on Windows"

File: packages/image_picker/image_picker_windows/test/image_picker_windows_test.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_windows/test/image_picker_windows_test.dart:57]: You are testing implementation details of `file_selector` here; these tests would all break if FileSelectorWindows changed its implementation (e.g., switched to a Dart implementation, or to using Pigeon).

Even though you can't use `FileSelectorPlatform.instance` in the implementation, you can still write the implementation against a `FileSelectorPlatform` variable, and just manually instantiate it with a `FileSelectorWindows`. Then you can make that instance settable by tests, and insert a mock implementation in these tests and validate against that mock. Then it doesn't matter how `FileSelectorWindows` works, because you'll just be verifying that you're using its API correctly.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/image_picker/image_picker_windows/test/image_picker_windows_test.dart
Comment (azchohfi) [reply packages/image_picker/image_picker_windows/test/image_picker_windows_test.dart]: @stuartmorgan I just pushed the changes. I believe that is what you means, which I agree that is a way better approach.

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (azchohfi) [reply packages/image_picker/image_picker_windows/lib/image_picker_windows.dart]: Done.

**Comment by azchohfi** [comment]
@stuartmorgan anything needed here?

File: packages/image_picker/image_picker_windows/test/image_picker_windows_test.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_windows/test/image_picker_windows_test.dart:102]: It would be good to add tests that passing `camera` as a source throws `UnimplementedError` for all the relevant methods.

**Review by stuartmorgan — APPROVED** [review]
One comment about adding a few more tests, but LGTM otherwise. The main thing this is waiting on is just the second review from @cbracken.

File: packages/image_picker/image_picker_windows/test/image_picker_windows_test.dart
Comment (azchohfi) [reply packages/image_picker/image_picker_windows/test/image_picker_windows_test.dart:126]: Done.

**Comment by azchohfi** [comment]
@cbracken anything else needed here?

**Comment by azchohfi** [comment]
Merged main again, just to keep this updated. Unlike the local_auth plugin, the image_picker plugin is already federated. @stuartmorgan / @cbracken what is keeping this PR from being reviewed/merged?

**Comment by stuartmorgan** [comment]
> @stuartmorgan / @cbracken what is keeping this PR from being reviewed/merged?

It's still @cbracken having time to review; Flutter policy requires two reviews for PRs from people outside the org before landing.

File: packages/image_picker/image_picker_windows/example/lib/main.dart
Comment (cbracken) [inline packages/image_picker/image_picker_windows/example/lib/main.dart:373]: Here and below: `controller` is nullable; is it safe to assume it's non-null in all these cases? If so, consider making it non-nullable which pushes the null-check up into `_previewVideo`, so no one needs to reason about nullability here.

File: packages/image_picker/image_picker_windows/example/lib/main.dart
Comment (cbracken) [inline packages/image_picker/image_picker_windows/example/lib/main.dart:199]: Optionally simplify down to:
```dart
Widget _handlePreview() => isVideo ? _previewVideo() : _previewImages();
```

File: packages/image_picker/image_picker_windows/example/lib/main.dart
Comment (cbracken) [inline packages/image_picker/image_picker_windows/example/lib/main.dart:55]: Consider making these private.

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (cbracken) [inline packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:20]: nit: here and below, add trailing periods on doc comments.

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (cbracken) [inline packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:49]: Since this is never reassigned, you can make this `static late final`.

Comment (cbracken) [inline packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:49]: Consider dropping `Instance` from the name; it doesn't add any information, since all variables are instances of some type. Plus it might save a line-wrap some day :)

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (cbracken) [inline packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:47]: This comment doesn't provide much info that isn't in the declaration itself. Consider documenting how this is used. Even something super short like:
```dart
/// The file selector used to prompt the user to select images or videos.
```

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#documentation-dartdocs-javadocs-etc

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (cbracken) [inline packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:56]: Here and below: Drop the 'Note that the'.

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-empty-prose

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (cbracken) [inline packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:117]: Is the intent to add other image sources in the future? If so, please file an issue to track this, add a `TODO` here and link the issue.

```
// TODO(username): Support ImageSource.foo, ImageSource.bar.
//                 See http://link/to/a/bug/123
```

If we will never support other image sources, throw `UnsupportedError` instead.

**Review by cbracken — CHANGES_REQUESTED** [review]
Overall looks good! Just a few comments.

File: packages/image_picker/image_picker_windows/example/lib/main.dart
Comment (azchohfi) [reply packages/image_picker/image_picker_windows/example/lib/main.dart:142]: I honestly don't know. I copied this example app from the other image_picker samples that are already in this repository. I'm not the original developer of it.

File: packages/image_picker/image_picker_windows/lib/image_picker_windows.dart
Comment (azchohfi) [reply packages/image_picker/image_picker_windows/lib/image_picker_windows.dart:49]: Can't be final since the tests need to replace this instance.

**Comment by azchohfi** [comment]
@cbracken let me know what you think.

File: packages/image_picker/image_picker_windows/example/lib/main.dart
Comment (cbracken) [reply packages/image_picker/image_picker_windows/example/lib/main.dart:152]: Ah makes sense. Ignore then, and we can do a separate pass over the lot of them once this has landed. Thanks for clarifying.

**Pull request merged**
